src/build-remote/build-remote.cc

```diff
@@ -123,11 +130,22 @@ static int _main(int argc, char * * argv)
                 for (auto & m : machines) {
                     debug("considering building on remote machine '%s'", m.storeUri);
 
-                    if (m.enabled && std::find(m.systemTypes.begin(),
-                            m.systemTypes.end(),
-                            neededSystem) != m.systemTypes.end() &&
-                        m.allSupported(requiredFeatures) &&
-                        m.mandatoryMet(requiredFeatures)) {
+                    if (!m.enabled) {
+                        debug(format("declined remote machine '%1%' because it is disabled") % m.storeUri);
+                    } else if (std::find(m.systemTypes.begin(), m.systemTypes.end(), neededSystem)
+                        == m.systemTypes.end()) {
+                        printInfo(format("declined remote machine '%1%' for building '%2%' because it does not have the needed system type: '%3%") % m.storeUri % drvstr % neededSystem);
+                    } else if (!m.allSupported(requiredFeatures)) {
+                        std::ostringstream joinedFeatures;
+                        for (const auto & i : requiredFeatures)
+                            joinedFeatures << i << ' '; // includes trailing space
+                        printInfo(format("declined remote machine '%1%' for building '%2%' because it does not have all required features: [ %3%]") % m.storeUri % drvstr % joinedFeatures.str());
+                    } else if (!m.mandatoryMet(requiredFeatures)) {
+                        std::ostringstream joinedFeatures;
+                        for (const auto & i : requiredFeatures)
+                            joinedFeatures << i << ' '; // includes trailing space
```

Comment:
Unless I'm missing something, there is a `concatStringsSep` function in `utils.hh` you could use for this?

Reply:
Done.

Reply:
I also noticed `concatStringsSep` does O(n²) string append, and `ostringstream` is no better; fixing it in PR #3941.